Implement dynamic OSv compiler

miha-plesko opened the pull request:
Added dynamic OSV compiler that uses forked Capstan tool
to create OSv unikernels. This Capstan is able to
set arbitrary logical image size during compose (as opposed
to default 10GB). Therefore a --size argument was added
to the 'unik build' command.
# Testing

You can use these three example applications to test the new funcionality:
[dynamic-compiler-examples.zip](https://github.com/emc-advanced-dev/unik/files/542248/dynamic-compiler-examples.zip)
Just navigate to project folder and run `./unik_openstack_up.sh` to create OSv unikernel, upload it to OpenStack and run it.

miha-plesko commented:
@ilackarms @gberginc could you please make a review on this PR? So the idea is, once we merge it, to then use this same forked Capstan for Java compiler. Because then user will be able to run it with arbitrary size, not just the default 10 gigs, as we've discussed some time ago.

ilackarms commented on pkg/client/images.go:
```diff
@@ -45,7 +45,7 @@ func (i *images) Get(id string) (*types.Image, error) {
 }
 
-func (i *images) Build(name, sourceTar, base, lang, provider, args string, mounts []string, force, noCleanup bool) (*types.Image, error) {
+func (i *images) Build(name, sourceTar, base, lang, provider, args string, mounts []string, force, noCleanup bool, size string) (*types.Image, error) {
```
I'm not a fan of changing the API for this feature. In rump, we dynamically calculate image size to use for the unikernel. We want the Unik api to remain as generalized as possible; we'd like to avoid adding in parameters that are specific to a single compiler. I don't think it should be much work to dynamically determine the minimal size a user image should be based on the size of the base image, plus whatever the user gives us to put on the disk.

miha-plesko replied:
Yes. But wouldn't it be great for user to be able to pick logical size on her own? Imagine VirtualBox where you specify maximum size for your .vdi HDD:
![capture](https://cloud.githubusercontent.com/assets/8102426/19600827/05c5bada-97a7-11e6-85ca-ab9d00058bcf.PNG)

I don't think this feature is special to this specific compiler. Any compiler could build image of arbitrary size, am I right? Of course, they don't do it yet, but they could.

Should you insist on removing this from the API, we could alternatively read this value from `manifest.yaml`. Like this:

``` yaml
# manifest.yaml
image_size: 10GB
```

Would you prefer us to go this way?

ilackarms commented on pkg/daemon/daemon.go:
```diff
@@ -492,6 +504,27 @@ func (d *UnikDaemon) initialize() {
 				mountPoints = strings.Split(mntStr, ",")
 			}
 
+			sizeStr := req.FormValue("size")
```
i'd like to see a dynamic determination of size, rather than bubbling this up to the user through changes to the API

miha-plesko replied:
See comment above - we can avoid bubbling by introducing `image_size` attribute into manifest.yaml.

One problem with dynamic determination is when you e.g. run database inside unikernel. If unikernel is barely big enough to fit libraries in it, where would the content of the database go?

ilackarms commented on pkg/daemon/daemon.go:
```diff
@@ -492,6 +504,27 @@ func (d *UnikDaemon) initialize() {
+			// If we are building qemu image for openstack, then we increase value of --size to match the
+			// closest flavor. Why? Because user expects image to be able to consume all the HDD space
+			// that flavor offers (when she runs it on OpenStack).
+			if providerName == openstack_provider && compilerName == compilers.OSV_DYNAMIC_OPENSTACK {
```
the daemon should be ignorant of provider-specific behavior

miha-plesko replied:
I agree, this is ugly. But I couldn't find another way to do it since the final size needs to be known prior compiler (Docker container) does it work.

ilackarms commented on pkg/daemon/daemon.go:
```diff
@@ -492,6 +504,27 @@ func (d *UnikDaemon) initialize() {
+			// closest flavor. Why? Because user expects image to be able to consume all the HDD space
+			// that flavor offers (when she runs it on OpenStack).
+			if providerName == openstack_provider && compilerName == compilers.OSV_DYNAMIC_OPENSTACK {
+				flavor, err := d.providers[providerName].(*openstack.OpenstackProvider).PickFlavor(int(size), 0)
```
i would suggest putting the 'pick a flavor with disk size >= some size X' logic into the openstack provider.Stage() itself, rather than preempting it in daemon

miha-plesko replied:
But `provider.Stage()` function is called after compiler builds the image. This is a problem, because image logical size can only be set during build.

ilackarms commented on pkg/providers/openstack/openstack_provider.go:
```diff
@@ -56,3 +58,13 @@ func (p *OpenstackProvider) newClientGlance() (*gophercloud.ServiceClient, error
+
+// PickFlavor is a wrapper of openstack.pickFlavor.
+// It is exported and autoperforms authentication step.
+func (p *OpenstackProvider) PickFlavor(diskMB int, memoryMB int) (*flavors.Flavor, error) {
```
provider-specific methods should be private

ilackarms commented on pkg/types/params.go:
```diff
@@ -30,6 +33,8 @@ type CompileImageParams struct {
 	Args       string
 	MntPoints  []string
 	NoCleanup  bool
+	// Size specifies logical size of the compiled image
```
same thing about changing the API

ilackarms commented on containers/compilers/osv/dynamic/Dockerfile:
```diff
@@ -0,0 +1,48 @@
+RUN mkdir /project_directory
+
+# Compose boot image and copy it to /project_directory (unik expects it there as a result)
+ENTRYPOINT cd /project_directory && \
```
why is this an ENTRYPOINT and not a CMD instruction?

miha-plesko replied:
I think ENTRYPOINT is to be used when you want command to remain unchanged when running container. This is the case here since UniK does not modify it (it passes all parameters through environment).

ilackarms replied:
CMD runs the command after it. ENTRYPOINT does the same, but allows arguments to be passed to that command. Since we never expect / want arguments to be passed here, the instruction should be CMD

miha-plesko replied:
Fixed

ilackarms commented on containers/compilers/osv/dynamic/Dockerfile:
```diff
@@ -0,0 +1,48 @@
+	git checkout develop && \
+	echo "Using forked capstan. Last commit was:" && \
+	git log -1 && \
+	godep restore && \
```
is godep restore necessary? it feels like we are adding unnecessary space to the image

gberginc commented on cmd/build.go:
```diff
@@ -124,4 +126,5 @@ func init() {
 	buildCmd.Flags().StringSliceVar(&mountPoints, "mountpoint", []string{}, "<string,repeated> specify up to 8 mount points for volumes")
 	buildCmd.Flags().BoolVar(&force, "force", false, "<bool, optional> force overwriting a previously existing")
 	buildCmd.Flags().BoolVar(&noCleanup, "no-cleanup", false, "<bool, optional> for debugging; do not clean up artifacts for images that fail to build")
+	buildCmd.Flags().StringVar(&imageSize, "size", "10GB", "<string,optional> logical image size i.e. by what size can it grow.")
```
Add ` `(space) in`<string,optional>`. Also rephrase the second part as I don't quite understand what`by what size can it grow.` means.

miha-plesko replied:
Done.

gberginc commented on containers/compilers/osv/dynamic/Dockerfile:
```diff
@@ -0,0 +1,48 @@
+FROM ubuntu:14.04
```
It looks like others are using mostly 16.04. Just checking if this version is still ok.

miha-plesko replied:
Switched to 16.04 and it seems to be working. Thanks for noticing!

gberginc commented on containers/compilers/osv/dynamic/Dockerfile:
```diff
@@ -0,0 +1,48 @@
+
+# Install prerequisites
+RUN apt-get update -y
+RUN apt-get install -y curl git qemu  # wget
```
Remove wget if it's not needed

miha-plesko replied:
Wooops 😄

gberginc commented on containers/compilers/osv/dynamic/Dockerfile:
```diff
@@ -0,0 +1,48 @@
+ENV PATH=$GOBIN:/usr/local/go/bin:$PATH
+
+# Build Capstan from source (use mikelangelo-project fork that supports package management)
+RUN go get github.com/tools/godep && \
```
Readability suggestion: align the following commands with the first line, for example:

```
RUN go get github.com/tools/godep && \
    go get github.com/cloudius-systems/capstan && \
```

Alternatively, add RUN in front of every command instead of `&& \` after each line.

miha-plesko replied:
It was a tab-vs-space issue here that made it appear alligned in my editor. Fixed by using spaces everywhere.

I think using single RUN with multiple commands has better performance over multiple RUNs, since each RUN creates a new layer to the container.

gberginc commented on containers/compilers/osv/dynamic/Dockerfile:
```diff
@@ -0,0 +1,48 @@
+	godep restore && \
+	go install
+
+
```
Remove one line

miha-plesko replied:
Done.

gberginc commented on containers/compilers/osv/dynamic/docker_files/root/.capstan/config.yaml:
```diff
@@ -0,0 +1,2 @@
+repo_url: https://mikelangelo-capstan.s3.amazonaws.com/
+disable_kvm: true
```
Please add the comment here to explain why we are disabling it.

Also, I see github commenting on the "no newline at the end of file" -- check other sources to see if they have newline and fix this in all the newly added files.

miha-plesko replied:
Done.

gberginc commented on containers/versions.json:
```diff
@@ -25,5 +25,6 @@
   "qemu-util": "6f5922561bbb86e3",
   "compilers-osv-java": "14f2183e5cb49482",
-  "compilers-mirage-ocaml-ukvm": "2c35c23771682e9a"
+  "compilers-osv-dynamic": "v0.0",
```
@ilackarms where do these hashes come from?

ilackarms replied:
from here: https://github.com/emc-advanced-dev/unik/blob/master/Makefile#L37
the Makefile tags containers with the docker image-id (which is some kind of checksum docker calculates whenever there are changes to the container). it doesn't actually matter what you put in here; when we merge any container changes, we'll build the container and push it locally, which will automatically update the version in `versions.json`

gberginc replied:
@ilackarms thanks for the clarification

gberginc commented on containers/compilers/osv/dynamic/docker_files/root/dot_capstan:
```diff
@@ -0,0 +1 @@
+.capstan
```
As we discussed, please remove this file from this PR.

miha-plesko replied:
Done.

gberginc commented on pkg/compilers/osv/osv_dynamic.go:
```diff
@@ -0,0 +1,43 @@
+	"github.com/emc-advanced-dev/unik/pkg/types"
+	unikutil "github.com/emc-advanced-dev/unik/pkg/util"
+	"os"
+	// "gopkg.in/yaml.v2"
```
Remove if not needed.

miha-plesko replied:
Done.

gberginc commented on pkg/daemon/daemon.go:
```diff
@@ -353,6 +353,18 @@ func NewUnikDaemon(config config.DaemonConfig) (*UnikDaemon, error) {
 	}
 
+	// osv dynamic
+	_compilers[compilers.OSV_DYNAMIC_QEMU] = &osv.OsvQemuCompiler{
```
I suggest to do `osvJavaQemuCompiler := &osv.OsvQemuCompiler...` and then use it in both cases. Similar to `osvJavaXenCompiler` in [L330](https://github.com/emc-advanced-dev/unik/pull/99/files#diff-06b532a2a5a5d3ad1d5341b3e00945efR330).

miha-plesko replied:
Done.

gberginc commented on pkg/daemon/daemon.go:
```diff
@@ -492,6 +504,27 @@ func (d *UnikDaemon) initialize() {
 				mountPoints = strings.Split(mntStr, ",")
 			}
 
+			sizeStr := req.FormValue("size")
```
Currently, the size is only used by OSv Qemu and OSv OpenStack provider. Would it make sense to check this and report a warning of `--size` is used in any other case? So, `size` would only be read in these two cases, otherwise a warning would be printed informing the user that the `size` has been ignored.

miha-plesko replied:
Good point. Resolved by implementing this functionality on the client-side (cmd/build.go) so that logrus prints to user's console not to daemon's.

gberginc commented on pkg/types/types.go:
```diff
@@ -133,11 +133,13 @@ const (
+	DefaultInstanceMemory int `json:"DefaultInstanceMemory"` //required for all compilers
+	// MinInstanceDiskMB specifies minimum disk requirement to run the instance.
+	// Used e.g. to pick flavor in the case of OpenStack provider.
+	MinInstanceDiskMB  int                `json:"MinInstanceDiskMB"`
```
Any way to change this so that the alignment of the types would not be changed?

miha-plesko replied:
I think not. Unless we make one-line comment add add it to the right of the attribute.

gberginc reviewed (commented):
First pass review done.

miha-plesko reviewed (commented):
Thank you @gberginc, I've applied most of the comments and repushed.

miha-plesko reviewed (commented):
Thank you for the review, @ilackarms. I understand that you prefer not to change the API, and that's fine with me. We can include size as part of manifest.yaml. 

I would not disable it completely because I cannot imagine how would UniK automatically know how much user-data will be provided when unikernel is running. Imagine Cassandra database.

I'm looking forward to discuss this further! :D

Commit:
Implement dynamic OSv compiler

Added dynamic OSV compiler that uses forked Capstan tool
to create OSv unikernels. This Capstan is able to
set arbitrary logical image size during compose (as opposed
to default 10GB) therefore we introduce "image_size" parameter
to the manifest.yaml.

Signed-off-by: Miha Pleško <miha.plesko@xlab.si>

miha-plesko commented:
So I've modified this PR according to the discussion above:
* removed --size argument for the `unik build` command to prevent core modifications of UniK
* supported `image_size` parameter in OSv's manifest yaml

Also, I've added a compiler for the NodeJS that understands same manifest for NodeJS than rump. Same example application works for Rumprun and OSv:
```
$ cd docs/examples/example-nodejs-app
$ unik build --name imgExample --path ./ --base osv --language node --provider openstack
$ unik run --instanceName instExample --imageName imgExample
```

@ilackarms, I'm looking forward to hear your opinion :)

ilackarms commented:
Following the convention that we use with Rumprun, the size of the boot volume should be dynamically calculated (based on the size of the kernel + all files the user put in the project directory + an extra 20mb for grub). If the user wants additional storage for data, that's what data volumes are intended for (which you create by specifying a `--size` flag). I know this doesn't fit all use cases (some apps may want everything on a single volume), but in the case of rumprun, you can mount additional volumes (i believe up to 8) anywhere you like, so the application can just refer to the mount point without having to know that it belongs to a different piece of storage.

gberginc commented on docs/compilers/osv.md:
````diff
@@ -16,3 +19,86 @@ Compiling Java on the OSv platform requires the following parameters be met:
+    main_file: YOUR_MAIN_FILE.js
+    image_size: "10GB"  # optional, defaults to "10GB"
+    ```
+    where you replace `YOUR_MAIN_FILE.js` with the relative path to your main file from the root directory of your project.
````
@miha-plesko If I understand this correctly, this is actually exactly the same as if you were running node app in other unikernels that are already supported in UniK? If so, please mention this, as it is a readme.

miha-plesko replied:
Dropped NodeJS support from this PR

gberginc commented on docs/compilers/osv.md:
```diff
@@ -1,6 +1,9 @@
 UniK uses OSv as a platform for compiling Java to unikernels.
+Additionally, UniK has a built-in support to directly run application packages provided in
+[MIKELANGELO Project](https://www.mikelangelo-project.eu)'s
+[format](https://github.com/mikelangelo-project/capstan/blob/develop/Documentation/ApplicationManagement.md#package-management).
```
@miha-plesko I just opened the entire readme and noticed that there is a section on Java, which still relies on the version of the current compiler. I wonder if it would be better to have a separate readme or at least make sure that users understand that Java compiler is not currently using the MIKE project.

miha-plesko replied:
Good idea, I've created a new file `osv-mikelangelo.md` besides original `osv.md` and moved dynamic compiler documentation there.

gberginc commented on pkg/daemon/daemon.go:
```diff
@@ -353,6 +353,24 @@ func NewUnikDaemon(config config.DaemonConfig) (*UnikDaemon, error) {
 		compilers: _compilers,
 	}
 
+	// osv dynamic
```
I think this should be added above L350.

miha-plesko replied:
I agree, fixed.

gberginc commented on pkg/os/device.go:
```diff
@@ -96,3 +102,29 @@ func IsExists(f string) bool {
+// If sizeStr contains number without unit (e.g. "150"), then MB is assumed.
+func ParseSize(sizeStr string) (MegaBytes, error) {
+	// assume MB if no unit is passed
+	rNoUnit, _ := regexp.Compile("([0-9]+)$")
```
I suggest to change the pattern to `"^[0-9]+$"`. Two changes:
- use of `^` at the beginning will require that the text starts with a number (it only contains a number)
- there is no selector in the regex `()` - you don't actually need to capture any data at this point. This also requires that you change the check in the next line as in case of a match, you will get just a single element, not two.

I see that other regexes will also allow for such texts which I think should be fixed.

miha-plesko replied:
Thanks. I've applied this and moved the 'append MB to the string'  functionality to better place

gberginc reviewed (commented):
@miha-plesko sorry for making these reviews rather sporadic. I think it may be easier if parts of this PR would be split into multiple. I'll think about proposing the split - currently, I think at least the Node support should be in a separate one. It definitely depends on this one, but it is not related to this one here.

@ilackarms what do you think? Or even better: what would be your preferred way to work on these changes?

ilackarms commented on pkg/compilers/osv/osv_node.go:
```diff
@@ -0,0 +1,143 @@
+	if err := yaml.Unmarshal(data, &config); err != nil {
+		return "", errors.New("failed to parse yaml manifest.yaml file", err)
+	}
+	if err := generateCapstanMetaFilesNode(params.SourcesDir, &config); err != nil {
```
since we're generating some files here that aren't necessary after this function returns, we should `defer` cleaning them up here (except in the case that CompileParams.NoCleanup == true)

miha-plesko replied:
Dropped NodeJS support out of this PR

miha-plesko reviewed (commented):
Dear @ilackarms, dear @gberginc, I've finally rebased and applied your comments. Apologies for this taking me so long, thank you for not closing the PR in the meantime :)

The most notable change that I did is that I've removed one of the two commits from this PR. The NodeJS branding of Dynamic compiler can be done later with a new PR as suggested by @gberginc.

@ilackarms Please let me know of your opinion on this PR. I remember that we spoke about trying to unify interface of this dynamic compiler with UniK's manifest.yaml, and I think it was only because we weren't very clear about how the forked Capstan works. Is there anything else we should discuss/modify regarding this?

miha-plesko commented:
Closing this PR temporarily until we get dependent PR's merged: [#111](https://github.com/emc-advanced-dev/unik/pull/111). and [#110](https://github.com/emc-advanced-dev/unik/pull/110).
